lib/engine/step/g_1848/dutch_auction.rb

```diff
@@ -0,0 +1,142 @@
+        # if any privates are left and everyone passes in a row, owned privates pay and auction continues
+        # if someone has bought a private and everything is minimum, you can pass
+
+        ACTIONS = %w[bid reduce].freeze
+        ACTIONS_WITH_PASS = %w[bid reduce pass].freeze
+
+        MIN_PRICES = {
+          'P1' => 0,
+          'P2' => 40,
+          'P3' => 80,
+          'P4' => 140,
+          'P5' => 140,
+          'P6' => 200,
+        }.freeze
+
+        def description
+          "Buy a Company or Reduce its Price by #{@game.format_currency(5)}"
+        end
+
+        def available
+          @companies
+        end
+
+        def auctioneer?
+          false
+        end
+
+        def committed_cash(_player, _show_hidden = false)
+          0
+        end
+
+        def pass_description
+          'Pass'
+        end
+
+        def process_bid(action)
+          company = action.company
+          price = company.min_bid
+          buy_company(current_entity, company, price)
+          @round.next_entity_index!
+        end
+
+        def process_pass(action)
+          entity = action.entity
+          @log << "#{entity.name} passes"
+          entity.pass!
+          all_passed! if entities.all?(&:passed?)
+          @round.next_entity_index!
+        end
+
+        def process_reduce(action)
+          action.entity.unpass!
+          company = action.company
+          company.discount += 5
+          price = company.min_bid
+          @log << "#{current_entity.name} reduces #{company.name} by £5 to #{@game.format_currency(price)}"
+          @round.next_entity_index!
+        end
+
+        def may_reduce?(company)
+          # Each private can be discounted a maximum of 6 times
+          company.min_bid > MIN_PRICES[company.sym]
+        end
+
+        def actions(entity)
+          return [] if @companies.empty?
+          return [] unless entity.player?
+
+          actions = entity.player.companies.empty? ? ACTIONS : ACTIONS_WITH_PASS
+
+          entity == current_entity ? actions : []
+        end
+
+        def setup
+          setup_auction
+          @companies = @game.companies.sort_by(&:min_bid)
+          @cheapest = @companies.first
+        end
+
+        def min_bid(company)
+          return unless company
+
+          company.value - company.discount
+        end
+
+        def may_purchase?(_company)
+          true
+        end
+
+        protected
+
+        def active_auction
+          false
+        end
+
+        def all_passed!
+          # Run a fake OR where all privates pay once
+          @game.payout_companies
+          @game.or_set_finished
+
+          entities.each(&:unpass!)
+        end
+
+        def buy_company(player, company, price)
+          company.owner = player
+          player.companies << company
+          # P1 can be reduced to free, so we disable `check_positive`
```

Comment:
why don't you only do it when it's p1 then

Reply:
You mean company.sym == 'P1' ... spend(..., check_positive=false)
else: spend(...)
?

Reply:
or spend(price, @game,bank, check_positive: company.sym != 'P1')

Reply:
780735bd